### PR TITLE
Change the default value of status variables to an error

### PR DESCRIPTION
**Pull request by AndrzejKurek**
Update a couple of status variables to start with an error value.

### CODE REVIEW

**Comment by AndrzejKurek** [comment]
Added changes to CCM module too.

**Comment by shelib01** [comment]
@AndrzejKurek , waht about mbedtls_entropy_add_source() function(line 141) ? Nobody use it?

**Comment by AndrzejKurek** [comment]
> @AndrzejKurek , waht about mbedtls_entropy_add_source() function(line 141) ? Nobody use it?

This function has a flow that changes ret only on error (and there are no writes to ret on success), so I figured that changing it to alter the normal flow is out of scope. I can change it though, if there's no other task for analysing the default flow of functions (but more changes like that should be probably introduced).

File: library/ssl_tls.c [hunk 1]
```diff
@@ -175,7 +175,7 @@ int mbedtls_ssl_check_record( mbedtls_ssl_context const *ssl,
                               unsigned char *buf,
                               size_t buflen )
 {
-    int ret = 0;
+    int ret = MBEDTLS_ERR_SSL_INTERNAL_ERROR;
```
Comment (piotr-now) [inline library/ssl_tls.c:178]: Do I see well that for `MBEDTLS_SSL_PROTO_TLS` and without` MBEDTLS_SSL_PROTO_DTLS` this function will never return success?

File: library/ssl_srv.c [hunk 2]
```diff
@@ -4213,7 +4213,7 @@ static int ssl_in_client_key_exchange_parse( mbedtls_ssl_context *ssl,
                                           unsigned char *buf,
                                           size_t buflen )
 {
-    int ret;
+    int ret = MBEDTLS_ERR_SSL_INTERNAL_ERROR;
```
Comment (piotr-now) [inline library/ssl_srv.c:4216]: Why did you choose `INTERNAL_ERROR` and not` FAULT_ERROR`?

File: library/ssl_srv.c [hunk 3]
```diff
@@ -4735,7 +4735,7 @@ static int ssl_write_new_session_ticket( mbedtls_ssl_context *ssl )
  */
 int mbedtls_ssl_handshake_server_step( mbedtls_ssl_context *ssl )
 {
-    int ret = 0;
+    int ret = MBEDTLS_ERR_SSL_INTERNAL_ERROR;
```
Comment (piotr-now) [inline library/ssl_srv.c:4738]: Question as above.

File: library/ssl_srv.c (hunk 2 above)
Comment (AndrzejKurek) [reply library/ssl_srv.c:4216]: This was the error that was used in other FI countermeasures for ssl as well.

File: library/ssl_tls.c (hunk 1 above)
Comment (AndrzejKurek) [reply library/ssl_tls.c:178]: Yes, as records are not used for TLS. See comments at line 183.

File: library/pk.c [hunk 4]
```diff
@@ -577,7 +577,7 @@ static int uecc_eckey_verify_wrap( void *ctx, mbedtls_md_type_t md_alg,
                        const unsigned char *sig, size_t sig_len )
 {
     int ret;
```
[the remark this answers is not in the document]
Comment (AndrzejKurek) [reply library/pk.c:579]: There is no flow that will result in this verification function returning success if ret is zero. If `extract_ecdsa_sig` is skipped by a fault injection - `uECC_verify` will fail.

File: library/ssl_srv.c (hunk 2 above)
Comment (AndrzejKurek) [reply library/ssl_srv.c:4216]: After another inspection, INTERNAL_ERROR was later refactored to `MBEDTLS_ERR_PLATFORM_FAULT_DETECTED` - you're right :) I'll change that.

**Comment by shelib01** [comment]
There are 2 additional functions in pk.c with not initialized  return code and function return  0 
https://github.com/ARMmbed/mbedtls/blob/d540daa2a69926e4a518ae76d907c21ba9fb0622/library/pk.c#L552
https://github.com/ARMmbed/mbedtls/blob/d540daa2a69926e4a518ae76d907c21ba9fb0622/library/pk.c#L677

**Comment by shelib01** [comment]
@AndrzejKurek , general question related to tls_cli.c, tls_srv.c and tls_ssl.c: it is about 20k lines of code. How did you check which functions should be changed ?

**Comment by shelib01** [comment]
https://github.com/ARMmbed/mbedtls/blob/d540daa2a69926e4a518ae76d907c21ba9fb0622/tinycrypt/ecc_dh.c#L165
Any reason this function was skiped ?

**Comment by shelib01** [comment]
In https://github.com/AndrzejKurek/mbedtls/blob/fi_error_codes/tinycrypt/ecc_dsa.c there are 3 functions with uninitialized error codes

**Comment by AndrzejKurek** [comment]
> There are 2 additional functions in pk.c with not initialized return code and function return 0
> https://github.com/ARMmbed/mbedtls/blob/d540daa2a69926e4a518ae76d907c21ba9fb0622/library/pk.c#L552
> 
> 
> https://github.com/ARMmbed/mbedtls/blob/d540daa2a69926e4a518ae76d907c21ba9fb0622/library/pk.c#L677
Are you sure you mean `uecc_eckey_check_pair`? It doesn't have a return value at all, and I dont think that returning 0 conditionally, if a check succeeds, is bad practice?
As for `pk_ecdsa_sig_asn1_from_uecc` - `ret` is used to check if `asn1_write_mpibuf` has written anything, it's not a return value from this function. We could assign 0 to it after `MBEDTLS_ASN1_CHK_ADD` but then it wouldn't differ from `return( 0 )` at all. Also, this function only transcodes signature formats, so causing a fault here won't result in a passed test elsewhere - it might only make more things fail.

**Comment by AndrzejKurek** [comment]
> In https://github.com/AndrzejKurek/mbedtls/blob/fi_error_codes/tinycrypt/ecc_dsa.c there are 3 functions with uninitialized error codes

In `uECC_sign_with_k` and `uECC_sign` - `r` is used to check for return values from called functions, not to end a natural flow. There's no way to return `r` without making sure that its value is `UECC_FAULT_DETECTED`, or, in case of `uECC_sign_with_k` - making sure that it's different from `UECC_SUCCESS`.

File: tinycrypt/ecc.c [hunk 5]
```diff
@@ -1224,7 +1224,7 @@ int uECC_valid_public_key(const uint8_t *public_key)
 
 int uECC_compute_public_key(const uint8_t *private_key, uint8_t *public_key)
 {
-	int ret;
+	int ret = UECC_FAULT_DETECTED;
```
[the remark this answers is not in the document]
Comment (AndrzejKurek) [reply tinycrypt/ecc.c:1227]: Good catch.

File: library/ssl_cli.c [hunk 6]
```diff
@@ -682,7 +682,7 @@ static void ssl_write_alpn_ext( mbedtls_ssl_context *ssl,
  */
 static int ssl_generate_random( mbedtls_ssl_context *ssl )
 {
-    int ret;
+    int ret = MBEDTLS_ERR_SSL_INTERNAL_ERROR;
```
[the remark this answers is not in the document]
Comment (AndrzejKurek) [reply library/ssl_cli.c:685]: This does not return 0 in default flow, and returning it in a nested condition is fine, if I'm not mistaken.

**Comment by AndrzejKurek** [comment]
Updated the PR with suggested changes. 

> @AndrzejKurek , waht about mbedtls_entropy_add_source() function(line 141) ? Nobody use it?

After analysing it further - I'm not sure if the solution I proposed is different from the previous behaviour that much. The only thing that changed is a potential place to jump to - previously it was the end of the function, now it's `ret = 0;`.

File: library/ssl_srv.c (hunk 2 above)
Comment (AndrzejKurek) [reply library/ssl_srv.c]: Fixed.

File: library/ssl_srv.c (hunk 3 above)
Comment (AndrzejKurek) [reply library/ssl_srv.c]: Fixed.

File: library/ccm.c [hunk 7]
```diff
@@ -74,7 +75,7 @@ int mbedtls_ccm_setkey( mbedtls_ccm_context *ctx,
                         const unsigned char *key,
                         unsigned int keybits )
 {
-    int ret;
+    int ret = MBEDTLS_ERR_PLATFORM_FAULT_DETECTED;
```
[the remark this answers is not in the document]
Comment (AndrzejKurek) [reply library/ccm.c:78]: Fixed.

File: library/ssl_cli.c [hunk 8]
```diff
@@ -3362,7 +3362,7 @@ static int ssl_parse_certificate_request( mbedtls_ssl_context *ssl )
 
 static int ssl_parse_server_hello_done( mbedtls_ssl_context *ssl )
 {
-    int ret;
+    int ret = MBEDTLS_ERR_SSL_INTERNAL_ERROR;
```
[the remark this answers is not in the document]
Comment (AndrzejKurek) [reply library/ssl_cli.c]: Fixed.

**Comment by shelib01** [comment]
> > In https://github.com/AndrzejKurek/mbedtls/blob/fi_error_codes/tinycrypt/ecc_dsa.c there are 3 functions with uninitialized error codes
> 
> In `uECC_sign_with_k` and `uECC_sign` - `r` is used to check for return values from called functions, not to end a natural flow. There's no way to return `r` without making sure that its value is `UECC_FAULT_DETECTED`, or, in case of `uECC_sign_with_k` - making sure that it's different from `UECC_SUCCESS`.

@AndrzejKurek this is exactly the problem. We return error  hardcoded success error code that not related to operation that function performed. It is very easy to jump to end of function, without perform any signature and return 0

**Comment by AndrzejKurek** [comment]
I have addressed some additional comments about `pk.c` and `ecc_dsa.c`.

**Comment by AndrzejKurek** [comment]
This PR is ready for a re-review.

File: library/pk.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (piotr-now) [reply library/pk.c:579]: What about double synchronized FI? If we prevent double synchronized FI with a second check after random_delay for ret_fi why we do not do that here.
I do not think that random_delay is necessary here but if we have ret set with MBEDTLS_ERR_PLATFORM_FAULT_DETECTED at the beginning we would have covered double synchronized FI for `extract_ecdsa_sig()` and `if( ret != 0 )`

[the remark this answers is not in the document]
Comment (AndrzejKurek) [reply library/pk.c]: Added `int ret = MBEDTLS_ERR_PLATFORM_FAULT_DETECTED;`.

**Comment by AndrzejKurek** [comment]
Added one more line in `pk.c`, 579:
`int ret = MBEDTLS_ERR_PLATFORM_FAULT_DETECTED;`.
Ready for re-review.

**Pull request merged**